### PR TITLE
feat(EXC-1716): Canister Backtrace on Trap

### PR DESCRIPTION
**Pull request by adambratschikaye**
Resolves EXC-1716. We introduce an optional field in the `Hypervisor::Trapped` error variant to contain a backtrace captured from the canister when it traps. The field will be `None` if the `canister_backtrace` feature is not implemented. Otherwise it will be `None` if the canister doesn't include a "name" custom section or the calling canister doesn't have permissions to view the backtrace.

Follow up changes will capture the backtrace in other error situations, dump it to the canister log, and strip it out of reject responses when the caller doesn't have sufficient permissions.

### CODE REVIEW

**Review by basvandijk — APPROVED** [review]

**Review by max-dfinity — APPROVED** [review]
Do you want to add your team as an owner or co-owner of the backtrace canister?

**Comment by adambratschikaye** [comment]
> Do you want to add your team as an owner or co-owner of the backtrace canister?

I think we are the owners of it, right? From this previous PR which introduced it: https://github.com/dfinity/ic/pull/1353/files#diff-3d36a1bf06148bc6ba1ce2ed3d19de32ea708d955fed212c0d27c536f0bd4da7R196

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -193,7 +193,7 @@ go_deps.bzl               @dfinity/idx
 /rs/replicated_state/src/page_map/                      @dfinity/ic-message-routing-owners @dfinity/execution
 /rs/rosetta-api/                                        @dfinity/finint
 /rs/rust_canisters/                                     @dfinity/nns-team
-/rs/rust_canisters/backtrace-canister                   @dfinity/execution
+/rs/rust_canisters/backtrace_canister                   @dfinity/execution
```
Comment (adambratschikaye) [inline .github/CODEOWNERS:196]: This was a typo in the original PR to create the subdirectory.

File: rs/embedders/src/wasm_utils/validation.rs [hunk 2]
```diff
@@ -1399,7 +1399,10 @@ pub fn wasmtime_validation_config(embedders_config: &EmbeddersConfig) -> wasmtim
     config.generate_address_map(false);
     // The signal handler uses Posix signals, not Mach ports on MacOS.
     config.macos_use_mach_ports(false);
-    config.wasm_backtrace(false);
+    config.wasm_backtrace(match embedders_config.feature_flags.canister_backtrace {
```
[the remark this answers is not in the document]
Comment (adambratschikaye) [reply rs/embedders/src/wasm_utils/validation.rs]: done

File: rs/embedders/src/wasmtime_embedder/system_api.rs [hunk 3]
```diff
@@ -1243,13 +1246,13 @@ pub(crate) fn syscalls<
             move |mut caller: Caller<'_, StoreData>, err_code: i32| -> Result<(), _> {
                 let err = match InternalErrorCode::from_i32(err_code) {
                     InternalErrorCode::HeapOutOfBounds => {
-                        HypervisorError::Trapped(TrapCode::HeapOutOfBounds)
+                        HypervisorError::Trapped {trap_code:TrapCode::HeapOutOfBounds, backtrace:None}
```
[the remark this answers is not in the document]
Comment (adambratschikaye) [reply rs/embedders/src/wasmtime_embedder/system_api.rs]: Yeah, the strings making long lines a few lines down were preventing cargo format from updating this. I split up the lines so not cargo format runs on this match block.

File: rs/embedders/tests/wasmtime_embedder.rs [hunk 4]
```diff
@@ -291,9 +291,10 @@ fn stack_overflow_traps() {
             assert_eq!(
                 result.err(),
                 Some(
-                    ic_interfaces::execution_environment::HypervisorError::Trapped(
-                        ic_interfaces::execution_environment::TrapCode::StackOverflow
-                    )
+                    ic_interfaces::execution_environment::HypervisorError::Trapped {
```
[the remark this answers is not in the document]
Comment (adambratschikaye) [reply rs/embedders/tests/wasmtime_embedder.rs]: Yeah it seems like there were a few places in this file where types were referenced by full paths that were already imported. I fixed this one and the the others as well.

File: rs/embedders/tests/wasmtime_embedder.rs [hunk 5]
```diff
@@ -291,9 +291,10 @@ fn stack_overflow_traps() {
...
+                        trap_code: ic_interfaces::execution_environment::TrapCode::StackOverflow,
```
[the remark this answers is not in the document]
Comment (adambratschikaye) [reply rs/embedders/tests/wasmtime_embedder.rs]: done

File: rs/execution_environment/tests/backtraces.rs [hunk 6]
```diff
@@ -1,15 +1,31 @@
...
+};
 use ic_registry_subnet_type::SubnetType;
-use ic_state_machine_tests::{StateMachine, StateMachineBuilder};
+use ic_state_machine_tests::{ErrorCode, StateMachine, StateMachineBuilder, StateMachineConfig};
 use ic_types::{CanisterId, Cycles};
 
 const B: u128 = 1_000 * 1_000 * 1_000;
 
-fn env_with_backtrace_canister() -> (StateMachine, CanisterId) {
+fn env_with_backtrace_canister(feature_enabled: bool) -> (StateMachine, CanisterId) {
```
[the remark this answers is not in the document]
Comment (adambratschikaye) [reply rs/execution_environment/tests/backtraces.rs]: done

File: rs/interfaces/src/execution_environment/errors.rs [hunk 7]
```diff
@@ -217,7 +241,24 @@ impl std::fmt::Display for HypervisorError {
             Self::InstrumentationFailed(err) => {
                 write!(f, "Could not instrument wasm module of canister: {}", err)
             }
-            Self::Trapped(code) => write!(f, "Canister trapped: {}", code),
+            Self::Trapped {
+                trap_code,
+                backtrace,
+            } => {
+                writeln!(f, "Canister trapped: {}", trap_code)?;
+                // When the wasm_backtrace feature is enabled, we can provide a
```
[the remark this answers is not in the document]
Comment (adambratschikaye) [reply rs/interfaces/src/execution_environment/errors.rs:250]: done

**Review by maksymar — APPROVED** [review]

**Pull request merged**